executor/oci/resolvconf.go

```diff
@@ -4,28 +4,45 @@ import (
 	"context"
 	"os"
 	"path/filepath"
+	"regexp"
+	"strings"
 
 	"github.com/docker/docker/libnetwork/resolvconf"
 	"github.com/docker/docker/pkg/idtools"
 	"github.com/moby/buildkit/util/flightcontrol"
 	"github.com/pkg/errors"
 )
 
-var g flightcontrol.Group[struct{}]
-var notFirstRun bool
-var lastNotEmpty bool
+// ipLocalhost is a regex pattern for IPv4 or IPv6 loopback range
+const ipLocalhost = `((127\.([0-9]{1,3}\.){2}[0-9]{1,3})|(::1)$)`
 
-// overridden by tests
-var resolvconfPath = resolvconf.Path
+var (
+	// Note: the default IPv4 & IPv6 resolvers are set to Google's Public DNS
+	defaultIPv4Dns = []string{"nameserver 8.8.8.8", "nameserver 8.8.4.4"}
+	defaultIPv6Dns = []string{"nameserver 2001:4860:4860::8888", "nameserver 2001:4860:4860::8844"}
+
+	localhostNSRegexp = regexp.MustCompile(`(?m)^nameserver\s+` + ipLocalhost + `\s*\n*`)
+
+	g            flightcontrol.Group[struct{}]
+	notFirstRun  bool
+	lastNotEmpty bool
+
+	// overridden by tests
+	resolvconfPath = resolvconf.Path
+)
 
 type DNSConfig struct {
 	Nameservers   []string
 	Options       []string
 	SearchDomains []string
 }
 
-func GetResolvConf(ctx context.Context, stateDir string, idmap *idtools.IdentityMapping, dns *DNSConfig) (string, error) {
+func GetResolvConf(ctx context.Context, stateDir string, idmap *idtools.IdentityMapping, dns *DNSConfig, hostNetwork bool) (string, error) {
 	p := filepath.Join(stateDir, "resolv.conf")
+	if hostNetwork {
+		p = filepath.Join(stateDir, "resolv-host.conf")
+	}
+
 	_, err := g.Do(ctx, stateDir, func(ctx context.Context) (struct{}, error) {
```

Comment:
This `Do` by `stateDir` does not look correct. If it gets called in parallel with different `dns/hostNetwork` setting there will be a collision and wrong value returned.